codegen-serde/src/main/kotlin/software/amazon/smithy/rust/codegen/serde/KotlinClientSerdeDecorator.kt

```diff
@@ -0,0 +1,654 @@
+    )
+
+class KotlinClientSerdeDecorator : ClientCodegenDecorator {
+    override val name: String = "ClientSerdeDecorator"
+    override val order: Byte = 0
+
+    override fun extras(
+        codegenContext: ClientCodegenContext,
+        rustCrate: RustCrate,
+    ) {
+        rustCrate.mergeFeature(SerdeFeature)
+        val generator = SerializeImplGenerator(codegenContext)
+        rustCrate.withModule(Module) {
+            serializationRoots(codegenContext).forEach {
+                generator.generateRootSerializerForShape(
+                    it,
+                )(this)
+            }
+            addDependency(SupportStructures.serializeRedacted().toSymbol())
+            addDependency(SupportStructures.serializeUnredacted().toSymbol())
+        }
+    }
+}
+
+/**
+ * All entry points for serialization in the service closure.
+ */
+fun serializationRoots(ctx: CodegenContext): List<Shape> {
+    val serviceShape = ctx.serviceShape
+    val walker = Walker(ctx.model)
+    return walker.walkShapes(serviceShape).filter { it.hasTrait<SerdeTrait>() }
+}
+
+class SerializeImplGenerator(private val codegenContext: CodegenContext) {
+    fun generateRootSerializerForShape(shape: Shape): Writable = serializerFn(shape, null)
+
+    /**
+     * Generates a serializer for a given shape. Collection serializer require using a wrapper structure. To handle this,
+     * `applyTo` allows passing in a `Writable` referring to an input. This returns a writable that wraps the input if necessary,
+     * e.g.
+     *
+     * `&self.map_shape` => `SerializeSomeSpecificMap(&self.map_shape)`
+     *
+     * Even if no wrapping is required, the writable that is returned includes the required serializer as dependencies.
+     */
+    private fun serializerFn(
+        shape: Shape,
+        applyTo: Writable?,
+    ): Writable {
+        val name = codegenContext.symbolProvider.shapeFunctionName(codegenContext.serviceShape, shape) + "_serde"
+        val deps =
+            when (shape) {
+                is StructureShape -> RuntimeType.forInlineFun(name, Module, structSerdeImpl(shape))
+                is UnionShape -> RuntimeType.forInlineFun(name, Module, serializeUnionImpl(shape))
+                is TimestampShape -> serializeDateTime(shape)
+                is StringShape, is NumberShape -> directSerde(shape)
+                is DocumentShape -> serializeDocument(shape)
+                else -> null
+            }
+        return writable {
+            val wrapper =
+                when {
+                    deps != null -> null
+                    shape is MapShape -> serializeMap(shape)
+                    shape is CollectionShape -> serializeList(shape)
+                    // Need to figure out the best default here.
+                    else -> serializeWithTodo(shape)
+                }
+            if (wrapper != null && applyTo != null) {
+                rustTemplate("&#{wrapper}(#{applyTo})", "wrapper" to wrapper, "applyTo" to applyTo)
```

Comment:
I think you can also do this:
```suggestion
                rust("&#1T(#2T)", wrapper, applyTo)
```

Reply:
yeah you can (this was actually originally the only way!). I don't necessarily think its clearer (but if it's clearer to you, that's certainly enough reason to change it.)